### PR TITLE
Fixed Validation in API

### LINKED ISSUE #3299: unexpected required data in Update patient [fhir]
### Describe the bug

https://github.com/openemr/openemr/pull/3298#issuecomment-606736136

cc @prondubuisi

### PR DESCRIPTION
**Pull request by yashrajbothra**
Signed-off-by: Yash Bothra <yashrajbothra786@gmail.com>

Fixes #3299

### CODE REVIEW

**Comment by yashrajbothra** [comment]
@sjpadgett Sir, Please suggest something better. Not so confident with this one

**Comment by yashrajbothra** [comment]
Hello Everyone, While trying to improve RestController I noticed there is much code duplicacy in both Controllers. So why don't we make some base class for Controllers and Services?

**Comment by sjpadgett** [comment]
> So why don't we make some base class for Controllers and Services?

This would certainly be appropriate. Why not work this while I look over this PR. I have some concerns I need to look into. Okay?

**Comment by yashrajbothra** [comment]
> > So why don't we make some base class for Controllers and Services?
> 
> This would certainly be appropriate. Why not work this while I look over this PR. I have some concerns I need to look into. Okay?

Did you mean I should start with base class and you will review this later?

**Comment by sjpadgett** [comment]
I just saying i'm looking closely at this and will take me longer to get a good review done. I hope to have today but maybe not as I have other projects i'm trying to kick off.

**Comment by yashrajbothra** [comment]
No Problem 👍 . I will first need to look up and find out how we can do this in the best manner. Can we use some library or framework for this or Pure PHP would be better choice?

**Comment by sjpadgett** [comment]
> Can we use some library or framework for this or Pure PHP would be better choice?

Since this is an API and not an App i'm thinking any base classes would be fairly straight forward. I think we can keep this to pure PHP pattern. However, i'm open on the matter.

**Comment by yashrajbothra** [comment]
> > Can we use some library or framework for this or Pure PHP would be better choice?
> 
> Since this is an API and not an App i'm thinking any base classes would be fairly straight forward. I think we can keep this to pure PHP pattern. However, i'm open on the matter.

Okay, Let's keep it pure as Its an API.

**Comment by yashrajbothra** [comment]
But before moving forward and as It will affect some of the ongoing projects. I would like to know the opinion of others as well.

cc @prondubuisi @im-Amitto @bradymiller @adunsulag and Everybody who is interested in this topic.

**Comment by sjpadgett** [comment]
That's fine but, in that case, you'll need to define what you hope to achieve with the base class. Starting with a controller base class.

**Comment by yashrajbothra** [comment]
Ya Sure. As of now, I am not so sure so I will need to go through codebase then I will share my thoughts :)

File: src/Services/PatientService.php [hunk 1]
```diff
@@ -200,6 +206,10 @@ public function update($pid, $data)
         );
     }
 
+    public function insertById($pid,$data){
+        return updatePatientData($pid,$data,'true');
+    }
```
[the remark this answers is not in the document]
Comment (yashrajbothra) [reply src/Services/PatientService.php:211]: isn't this means that the user wants to create on particular PID? if user wants to create a user by certain user id then I think it will be required to pass PID

File: src/Services/PatientService.php [hunk 2]
```diff
@@ -35,14 +36,14 @@ public function __construct()
     {
     }
 
-    public function validate($patient)
+    public function validate($patient, $type)
```
[the remark this answers is not in the document]
Comment (yashrajbothra) [reply src/Services/PatientService.php:39]: Yup! that will be better 
Thanks 👍

**Comment by yashrajbothra** [comment]
> Overly complicated.
> For me a service should self validate in the appropriate part of the service if it is that important to the workflow. What if validate is not called by controller? This way service is not concerned with the passed in context. It's handled in the action.
> 
> Anyway, give it some more thought!

Should we call validate in service methods like insert, update etc so we don't need to call then from the controller and will be easily managed

File: src/Services/PatientService.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (sjpadgett) [reply src/Services/PatientService.php:221]: We don't want to allow create from a PUT! That's what insert is for plus pid is created from a sequence.
Also our patient insert should also be creating a blank insurance record. Just like a create patient workflow...

**Comment by sjpadgett** [comment]
>S hould we call validate in service methods like insert, update etc so we don't need to call then from the controller and will be easily managed

This would be my preference.  You'd need to handle failure from validate in controller on return though.

File: src/RestControllers/PatientRestController.php [hunk 3]
```diff
@@ -40,15 +40,21 @@ public function post($data)
...
-
+        if ($this->patientService->isPatient($pid)) {
+            $validationResult = $this->patientService->validate($data, 'update');
+            $action = 'update';
+            $status = 200;
+        } else {
+            $validationResult = $this->patientService->validate($data, 'insert');
+            $action = 'insertById';
+            $status = 201;
+        }
```
Comment (bradymiller) [inline src/RestControllers/PatientRestController.php:51]: hi @yashrajbothra , pretty sure that PUT should always be just "update"

File: src/Services/FHIR/FhirPatientService.php [hunk 4]
```diff
@@ -90,16 +115,21 @@ public function setId($id)
...
 
     public function insert($data)
     {
         return $this->patientService->insert($data);
     }
 
+    public function insertById($pid, $data)
+    {
+        return $this->patientService->insertById($pid, $data);
+    }
```
Comment (bradymiller) [inline src/Services/FHIR/FhirPatientService.php:131]: this function seems odd

File: src/Services/FHIR/FhirPatientService.php [hunk 5]
```diff
@@ -119,4 +149,9 @@ public function getAll($searchParam)
     {
         return $this->patientService->getAll($searchParam);
     }
+
+    public function isPatient($pid)
+    {
+        return $this->patientService->isPatient($pid);
+    }
```
Comment (bradymiller) [inline src/Services/FHIR/FhirPatientService.php:156]: this function seems odd

File: src/Services/PatientService.php [hunk 6]
```diff
@@ -14,6 +14,7 @@
 
 namespace OpenEMR\Services;
 
+require_once("{$GLOBALS['srcdir']}/patient.inc");
```
Comment (bradymiller) [inline src/Services/PatientService.php:17]: when bring in cases like this (in an autoloaded class), generally use "hard-coded" path:
```
require_once(dirname(__FILE__) . "/../../library/patient.inc");
```
And to make it clear, generally make a line of spacing between this and the `use` entries

File: src/Services/PatientService.php [hunk 7]
```diff
@@ -35,14 +36,24 @@ public function __construct()
...
+                $type = 'required';
+                break;
+            default:
+                $type = 'required';
+                break;
+        }
+        $validator->$type('fname')->lengthBetween(2, 255);
+        $validator->$type('lname')->lengthBetween(2, 255);
+        $validator->$type('sex')->lengthBetween(4, 30);
+        $validator->$type('DOB')->datetime('Y-m-d');
```
Comment (bradymiller) [inline src/Services/PatientService.php:57]: But isn't a pid required on the update?
Also note that somebody will likely be building this out to add more optional validations for the insert (and the update) and to play nicely with the current layout engine. Thus may not make sense to make it always optional for update and required for insert.

File: src/Services/PatientService.php [hunk 8]
```diff
@@ -58,6 +69,11 @@ public function getPid()
         return $this->pid;
     }
 
+    public function isPatient($pid)
+    {
+        return getPatientData($pid,'pid');
+    }
```
Comment (bradymiller) [inline src/Services/PatientService.php:75]: this functions seems odd

File: src/RestControllers/FHIR/FhirPatientRestController.php [hunk 9]
```diff
@@ -58,14 +58,22 @@ public function put($pid, $fhirJson)
...
-        $validationResult = $this->fhirPatientService->validate($data);
+        if ($this->fhirPatientService->isPatient($pid)) {
+            $validationResult = $this->fhirPatientService->validate($data, 'optional');
+            $action = 'update';
+            $status = 200;
+        } else {
+            $validationResult = $this->fhirPatientService->validate($data);
+            $action = 'insertById';
+            $status = 201;
+        }
```
Comment (bradymiller) [inline src/RestControllers/FHIR/FhirPatientRestController.php:69]: this stuff seems odd. shouldn't POST always be insert and PUT always be update?

Comment (yashrajbothra) [reply src/RestControllers/FHIR/FhirPatientRestController.php:69]: what if user adds a data with an id which doesn't exist in the system in such case as per FHIR guide they suggest to create one and give response with status 201 

https://www.hl7.org/fhir/http.html#update
they are using "MAY" here so we can ignore this ?

File: src/Services/PatientService.php (hunk 6 above)
Comment (yashrajbothra) [reply src/Services/PatientService.php:17]: Thanks! Will keep that in mind :)

File: src/RestControllers/FHIR/FhirPatientRestController.php (hunk 9 above)
Comment (bradymiller) [reply src/RestControllers/FHIR/FhirPatientRestController.php:69]: That seem a bit odd to offer though. For example, what if they do:
pid
mname

Just to change the mname. And now if the pid does not exist yet, OpenEMR will add:
1. A user with just a mname set
2. And directly setting the pid could be very problematic and break things.

Comment (yashrajbothra) [reply src/RestControllers/FHIR/FhirPatientRestController.php:69]: IMO this should be the job of patch and update will update the whole resource and if not found will create one.

PATCH will just patch some data of the existing user and never create a new resource.

Comment (yashrajbothra) [reply src/RestControllers/FHIR/FhirPatientRestController.php:69]: image

image

Comment (bradymiller) [reply src/RestControllers/FHIR/FhirPatientRestController.php:69]: I honestly didn't even know about the PUT/PATCH stuff. Just did a nice read here:
https://medium.com/backticks-tildes/restful-api-design-put-vs-patch-4a061aa3ed0b

If we wish to support PUT/PATCH correctly, looks like we'd need to change calls/routes a bit (I would not be against this and would defer to @sjpadgett and others such as yourself with REST expertise on this). We would have a tough time supporting some PUT calls (like for patient) since setting whatever pid you want can really break things. Of course when we migrate to uuid rather than id in tables, then this kind of thing will be much easier (in the super far distant future).

Comment (bradymiller) [reply src/RestControllers/FHIR/FhirPatientRestController.php:69]: A good first step will be adding PATCH here, so it even works on the dockers:
https://github.com/openemr/openemr-devops/blob/master/docker/openemr/5.0.3/openemr.conf#L25

Comment (yashrajbothra) [reply src/RestControllers/FHIR/FhirPatientRestController.php:69]: We can leave that create new user part from PUT for now as it is an optional thing to do.

**Comment by sjpadgett** [comment]
How many times do I have to say this, We are not going to support versions!!!! We will not be implementing the full REST interface!!!
Our gotcha with all of this is we don't have sql builders in our service to only build updates to what is passed in. Hard coding updates doesn't work...

**Comment by bradymiller** [comment]
@sjpadgett I'd like to hear that one more time (hard of hearing over here) :)

Seriously, though, sounds good and simplifies things.

File: src/RestControllers/FHIR/FhirPatientRestController.php (hunk 9 above)
Comment (prondubuisi) [reply src/RestControllers/FHIR/FhirPatientRestController.php:69]: > what if user adds a data with an id which doesn't exist in the system in such case as per FHIR guide they suggest to create one and give response with status 201
> 
> https://www.hl7.org/fhir/http.html#update
> they are using "MAY" here so we can ignore this ?

Hello @yashrajbothra with the current API's if one tries an update with an ID that doesn't exist, usually nothing is returned, and no update is done on the database. I agree it would be a nice to have to create a better response, which could work with a method that sees if IDs used across API's are valid and already existing. I do think you can break your Current PR into smaller PR's and of course create issues for each PR first.

**Comment by sjpadgett** [comment]
For us PATCH is an update. We only have one version of patient. We don't have the infrastructure to have versions of patient records or encounters and on and on.
Our service updates need to be reworked to build update sql to only update values passed in from the REST call. Also validate if pid is an existing pid else error.

**Comment by yashrajbothra** [comment]
After reading some blogs and rfc spec I got more confused. People on the web are telling just to use PATCH for partial update and PUT for full resource update. And the versions are only specified in FHIR guide and we can't implement that but what about standard API we using PUT there too.

TLDR; Just need to know why to use PUT for partial update and not PATCH.

RFC
https://tools.ietf.org/html/rfc5789

Some Discussions 
https://apisyouwonthate.com/blog/put-vs-patch-vs-json-patch
https://stackoverflow.com/questions/19732423/why-isnt-http-put-allowed-to-do-partial-updates-in-a-rest-api

File: src/RestControllers/FHIR/FhirPatientRestController.php (hunk 9 above)
Comment (adunsulag) [reply src/RestControllers/FHIR/FhirPatientRestController.php:69]: The majority of REST apis I've worked with rarely implement PUT/PATCH as the resources involved over GET/POST are more complicated and difficult to implement. 

I've never seen any REST implementation that lets PUT create a resource even though the RFC specifies it.  Just my 2 cents.  Creation of resources is supposed to be done via POST.

**Comment by sjpadgett** [comment]
> I'd like to hear that one more time (hard of hearing over here) :)

lol, I certainly didn't mean to be sarcastic but I woke for one of my hundred trips to the head last night and had to say my 2 cents:) I did sleep better afterwards though:)

@yashrajbothra  and @silvistershebin 
I'm very pleased to see ya's doing your research, being creative and most of all, defending your perspectives through citations and a little push back. These attributes will serve you well.

I've tried to be clear that with FHIR we simply have to temper what we can support. Yash, you're correct to point out how a normal FHIR server MAY handle a PUT that doesn't exist in the servers repository by doing an insert however, even there are restriction such as a custom id has to be provided. Now because a FHIR server normally tracks versions of updates where you may have several version of a patient record that you may want to patch or because the server may want to limit bandwidth, a patch may be preferred against receiving an entire resource data object.
For us, an update should be a patch or visa versa. Possibly handled through routes. I believe i've mentioned this.

Still we have to realize our current infrastructure and it's limitations. When we discuss a separate database for FHIR the main reason would be for setting up repositories that store entire resource objects with a structure that support versioning. But, it is just that, a repository. We'd still need to make the leap to openemr workflow. Our mission is not to be store.

Our mission priority is to support ONC 2015(or 2020) certification. ONC for a modular EHR API requirements are concerned with read actions. They, as I, understand that allowing updates/inserts in a EHR setting can be problematic. For us, we need to be careful with these types of transactions meaning, we'd need a review process before making patient data changes permanent. Unsolicited updates/insert should be avoided unless we have some better control for these changes.

Having said that, we should have the capability in our bag of tricks while we are doing this. Besides setting up validation for the different actions we are still faced with with how our services handle these storage actions on the backend.
Currently patient service is hard coded for a set of columns to be saved regardless of what the action request requires, possibly causing columns to be overwritten or erased that are not part of the update request. Big problem.
Normally we may reconsider current workflow by restructuring to doctrine and symfony using those tools for building table queries but that'd require considerable work and may be overkill. However, if we had all our infrastructure database/tables broke out into entities then that'd may be ideal but not something I want to take on:)

Ya need to take a look at patient service update and rework to build our sql to only updates passed into the action.

Hope this clarify some and it's fine to push back if you wish but, just not against me!:) kidding!

**Comment by yashrajbothra** [comment]
Thanks you Sir for explaining so well. IDK why I forget that line again n again.
> For us, an update should be a patch or visa versa. Possibly handled through routes. I believe i've mentioned this.

Does this mean we should just change the HTTP method PUT to PATCH in routes.

> Normally we may reconsider current workflow by restructuring to doctrine and symfony using those tools for building table queries but that'd require considerable work and may be overkill. However, if we had all our infrastructure database/tables broke out into entities then that'd may be ideal but not something I want to take on:)
> 
> Ya need to take a look at patient service update and rework to build our sql to only updates passed into the action.

Is your next concept PR going to affect these changes? If that then I could pause this for a while and work on some other PR?

File: src/Services/FHIR/FhirPatientService.php [hunk 10]
```diff
@@ -54,32 +54,59 @@ public function createPatientResource($resourceId = '', $data = '', $encode = tr
...
         }
-        $data["phone_contact"] = $phone["value"];
-        $data["DOB"] = $fhirJson["birthDate"];
-        $data["sex"] = ucwords($fhirJson["gender"]);
-
+        if (isset($fhirJson["telecom"])) {
+            $phone = [];
+            foreach ($fhirJson["telecom"] as $phone) {
+                if ($phone["use"] == "mobile") {
+                    $name = $phone;
```
Comment (sjpadgett) [inline src/Services/FHIR/FhirPatientService.php:96]: This doesn't make sense! $name not used. Could just go ahead and set phone here.
Just fix when we add the rest of the possible phones

**Comment by bradymiller** [comment]
hi @yashrajbothra and @sjpadgett , Code looks great to me; if it's working, then would bring it in.

Regarding the PATCH issue, looks like @sjpadgett brought in a PATCH route. So, sounds like we should then change All PUT routes to PATCH then, correct?
(note part of this PR, but if so, then we should make an issue for that)

File: src/Services/FHIR/FhirPatientService.php (hunk 10 above)
Comment (yashrajbothra) [reply src/Services/FHIR/FhirPatientService.php]: BTW I didn't like the way we parse our FHIR resources. I think we could do better.
Ohno, One more task in the bucket list. My Bucket List is growing exponentially. And I am too lazy to pull requests. Will make a better schedule from now😤

File: src/Services/FHIR/FhirPatientService.php [hunk 11]
```diff
@@ -54,32 +54,76 @@ public function createPatientResource($resourceId = '', $data = '', $encode = tr
...
+                            case 'home':
+                                $data["phone_home"] = $telecom["value"];
+                                break;
+                            case 'work':
+                                $data["phone_biz"] = $telecom["value"];
+                                break;
+                            default:
+                                $data["phone_contact"] = $telecom["value"];
+                                break;
+                        }
```
Comment (sjpadgett) [inline src/Services/FHIR/FhirPatientService.php:110]: very good.

**Comment by sjpadgett** [comment]
> Regarding the PATCH issue

We still need UPDATE as it will be goto. PATCH will likely not be used often and we treat as update anyway. Generally in FHIR an update would generate a new version of the record where patch simply doesn't. Since we don't track version of updates then become moot.
It'd be nice to track update versions especially of patient data but, would require some heady stuff left for the younger folks then me.....

File: src/Services/FHIR/FhirPatientService.php (hunk 11 above)
Comment (yashrajbothra) [reply src/Services/FHIR/FhirPatientService.php:110]: Will review once again after dinner.

Comment (sjpadgett) [reply src/Services/FHIR/FhirPatientService.php:110]: The way you have now is great. I was just commenting I likey:)

**Comment by bradymiller** [comment]
Regarding PATCH/UPDATE issue, lets step away from the versioning and FHIR issues. Goal is to just make things consistent (now that we have a PATCH route in the mix). Notably that PATCH route that was added has an identical PUT route that does the same exact thing:
https://github.com/openemr/openemr/blob/master/_rest_routes.inc.php#L372-L381

From my understanding for REST purposes:
1. PUT is supposed to create a new entry if one does not exist for the item that is being modified
2. PATCH will only modify the item if it exists
(this appears to be same expected behavior for FHIR detailed above by @sjpadgett )

So, does it make sense to change all the PUT requests to PATCH to be consistent? Or should we remove that lone PATCH? Or am I missing something obvious?

**Comment by yashrajbothra** [comment]
I think its optional to create a new resource in PUT on FHIR or in any API.
image

**Comment by bradymiller** [comment]
I guess what is truly confusing me here is why a PATCH route was added that has an identical PUT route that both do the same thing:
https://github.com/openemr/openemr/blob/master/_rest_routes.inc.php#L372-L381

**Comment by bradymiller** [comment]
I am gonna take a stab to answer my own question. Guessing FHIR server needs to support PATCH requests. So, even though PUT/PATCH are treated the same on our FHIR implementation, we still need to support the request. Thus all the FHIR PUT routes will have an identical PATCH route? (thus, not an issue for the standard API)

**Comment by yashrajbothra** [comment]
> I am gonna take a stab to answer my own question. Guessing FHIR server needs to support PATCH requests. So, even though PUT/PATCH are treated the same on our FHIR implementation, we still need to support the request. Thus all the FHIR PUT routes will have an identical PATCH route? (thus, not an issue for the standard API)

PUT : Updated/Replace the resource.
PATCH : Update the give data and leave rest of the data as is.

But as per my understaning FHIR or any API PUT request, in general, should update the whole resource and not partial update. That's why they have huge difference. And  if we decide to do I think we are not following REST rules 😕

File: src/Services/PatientService.php [hunk 12]
```diff
@@ -38,44 +39,48 @@ public function __construct()
...
+        
+        $validator->context('insert', function (Validator $context) {
+            $context->required('first_name')->lengthBetween(2, 30);
+            $context->required('fname')->lengthBetween(2, 255);
+            $context->required('lname')->lengthBetween(2, 255);
+            $context->required('sex')->lengthBetween(4, 30);
+            $context->required('DOB')->datetime('Y-m-d');
+        });
+        
+        $validator->context('update', function (Validator $context) {
```
Comment (yashrajbothra) [inline src/Services/PatientService.php:60]: http://validator.particle-php.com/en/latest/contexts/
I wanted to apply the example given there as "Extended example of copying".
I tried but didn't work for me I think its bug in docs or in Validator.
Can you please help me ?

**Comment by sjpadgett** [comment]
> I am gonna take a stab to answer my own question. Guessing FHIR server needs to support PATCH requests. So, even though PUT/PATCH are treated the same on our FHIR implementation, we still need to support the request. Thus all the FHIR PUT routes will have an identical PATCH route? (thus, not an issue for the standard API)

Yep! If we support patch then this is how we do it. We don't have to implement anything we don't want. That's what capability statements are for. The only reason I added a patch route was for example. I don't know where this notion comes from that to do an update, client has to send an entire record. In FHIR world, an update is about versions of a record, not whether an update contains the entire possible record content.
For our standard API I get we need to support updates for purposes of using in core. FHIR on the other hand, this is not the case. We are an API not a FHIR server. What a client is allowed to access is entirely up to us as long as the data we serve follows the spec.

File: src/Services/PatientService.php (hunk 12 above)
Comment (yashrajbothra) [reply src/Services/PatientService.php:60]: Oops I think we can't use this it doesn't copy the validation just copy the context :(

**Comment by prondubuisi** [comment]
@bradymiller Creating a user, when one does not exist might work for FHIR, but does not look good for APIs. Trying an update with a non existing resource should return an error message for API's not create the resource. Don't see any use case where that works for API's. FHIR and API's obviously have different needs.

**Comment by sjpadgett** [comment]
> Creating a user, when one does not exist

we are not doing that! ever for standard or FHIR. Lets move on...

**Comment by bradymiller** [comment]
yep, there has been some confusion with terms being used (PUT, PATCH, update). At the end of the day, will continue with the common sense route as we are doing (@prondubuisi , this plan is in line with your post) :)

**Comment by prondubuisi** [comment]
No issues @bradymiller @sjpadgett lets keep moving fast and breaking things!

File: src/Services/PatientService.php [hunk 13]
```diff
@@ -52,16 +51,16 @@ public function validate($patient, $context, $id = null)
...
+            $context->required('lname', 'Last Name')->lengthBetween(2, 255);
+            $context->required('sex', 'Gender')->lengthBetween(4, 30);
+            $context->required('DOB', 'Date of Birth')->datetime('Y-m-d');
         });
         
         $this->validator->context('update', function (Validator $context) {
             $context->copyContext('insert', function ($rules) {
                 foreach ($rules as $key => $chain) {
-                    $this->validator->optional($key);
+                    $chain->required(false);
```
Comment (yashrajbothra) [inline src/Services/PatientService.php]: Not so confident about this. 
Lets see if https://github.com/particle-php/Validator/pull/195 it gets merged 🤔

File: src/Services/PatientService.php [hunk 14]
```diff
@@ -52,16 +51,16 @@ public function validate($patient, $context, $id = null)
         
         $this->validator->context('insert', function (Validator $context) {
             $context->required('first_name')->lengthBetween(2, 30);
```
Comment (yashrajbothra) [inline src/Services/PatientService.php:53]: Whats diff btw `fname` and `first_name` 😕

**Comment by yashrajbothra** [comment]
Cleared my testing. Ready to merge 👍

File: src/Services/PatientService.php (hunk 14 above)
Comment (sjpadgett) [reply src/Services/PatientService.php:53]: There isn't a first_name in patient data. Generally that syntax is used for user etc.  So, we need to be careful that array pass ins follow table column naming conventions.
Probably need to fix this Yash....

**Comment by sjpadgett** [comment]
@yashrajbothra As soon as you fix first_name, i'll bring in.
So, what's your plan next? Do you want to do a challenging resource or knock out a couple minor ones? 
@silvistershebin @bradymiller 
I'm very much interested in getting our core resources done even if it means we develop initially for read ie Read/GET w/search. We can circle back to updates/puts/post etc.
For ONC FHIR this is required and I want us ready when v5.0.3 is released. Also, many client applications will be mainly concerned with portal like workflows where if an immediate need arises for their application to add chart data then, they can help contribute to API.

Again, many resources are referenced by other resources in a canonical ref for lookup/read. We need to also implement Document where CCDA generation and inclusion is done and not to forget Devices which @RachelEllison  is working towards for MU3 so some of that infrastructure is being worked.

So, though a complete API is desirable, the implementation of all the core resources for at least read is necessary for v5.0.3. Taking this route will, along the way, gain us experience and basis for the other actions(post etc).

https://build.fhir.org/ig/HL7/US-Core-R4/#us-core-profiles

@prondubuisi Are you going to be working FHIR?

**Comment by silvistershebin** [comment]
Hi @sjpadgett  Sir, I thought of working on the Immunization Resource or the Condition Resource. Will it be fine? Or else you can let me know if you want me to work on any other specific resource.

**Comment by sjpadgett** [comment]
> Immunization Resource or the Condition Resource

Either or both will be fine and thank you

File: src/Services/PatientService.php (hunk 14 above)
Comment (yashrajbothra) [reply src/Services/PatientService.php:53]: Do i just need to remove this ?

Comment (sjpadgett) [reply src/Services/PatientService.php:53]: As long as we still require fname for insert. ie first_name is fname etc

Comment (yashrajbothra) [reply src/Services/PatientService.php:53]: Ohh so we need to show a clear message that `first name` is required because `fname` isn't so clear. Right?

Comment (sjpadgett) [reply src/Services/PatientService.php:53]: depends how granular you want to report error Yash. I think first name is fine for both if fname data element not passed in or value is missing.

Comment (yashrajbothra) [reply src/Services/PatientService.php:53]: Ok for that we have already given the name of every index to Validator. That will be consistent and will do a better job.

Comment (sjpadgett) [reply src/Services/PatientService.php]: okay, lets fix and get this in so I can test while working on other stuff! fname is our guy...

Comment (yashrajbothra) [reply src/Services/PatientService.php]: Done 👍

**Comment by sjpadgett** [comment]
Thanks Yash. Good job.  I guess we finally got there:) I'm really interested to see what you come up with for Observation and looking forward to next review.
